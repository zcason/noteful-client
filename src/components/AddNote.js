import React, { Component } from 'react';
import SidebarSection from './sections/SidebarSection';
import DefaultContext from './context/DefaultContext';
import { withRouter } from 'react-router-dom';
import ErrorBoundary from '../errors/ErrorBoundary';
import PropTypes from 'prop-types';

class AddNote extends Component {
    static contextType = DefaultContext;
    handleOnSumbit = (form) => {
        const f = new FormData(form);
        const data = {
            name: f.get("noteName"),
            modified: new Date().toISOString(),
            content: f.get("noteContent"),
            folderId: f.get("folderId")
        }
        this.addNote(data);
    }
    addNote = (data) => {
        fetch(`${this.context.url}/notes/`, {
            method: 'POST',
            headers: {
              'content-type': 'application/json'
            },
            body: JSON.stringify(data)
        })
        .then( r=>{
            this.context.updateStore();
            this.props.history.push('/');
        })
        .catch( e => {
            throw new Error("Error creating note");
        } );
    }
    getFolderList = () => {
        if ( typeof this.props.store !== "object" || this.props.history.location.pathname.includes("/note/") ) return;

        return this.props.store.folders.map( (folder,i) => {
            return (
                <option
                    key={i}
                    value={folder.id}
                >
                    {folder.name}
                </option>
            );
        });
    }
    render() {
        return (
            <div className="App">
                <ErrorBoundary message="Sidebar Section Error">
                    <SidebarSection 
                        store={this.props.store}
                        history={this.props.history}
                    />
                </ErrorBoundary>
                <ErrorBoundary message="Main Section Error">
                    <main className="section--main">
                        <form
                            className="add--form"
                            onSubmit={(e)=>{
                                e.preventDefault();
                                this.handleOnSumbit(e.target);
                            }}
                        >   
                            <div className="add--form--field">
                                <label htmlFor="noteName">Name:</label>
                                <input type="text" id="noteName" name="noteName" required />
                            </div>
                            <div className="add--form--field">
                                <label htmlFor="noteContent">Content:</label>
                                <textarea id="noteContent" name="noteContent" required></textarea>
                            </div>
                            <div className="add--form--field">
                                <label htmlFor="folderId">Folder:</label>
                                <select name="folderId" id="folderId" required>
                                    {this.getFolderList()}
                                </select>
                            </div>
                            <div className="add--form--field">
                                <button type="submit">Submit</button>
                            </div>
                        </form>
                    </main>
                </ErrorBoundary>
            </div>
        )
    }
}

AddNote.propTypes = {
    store: PropTypes.object.isRequired
}

export default withRouter(AddNote);