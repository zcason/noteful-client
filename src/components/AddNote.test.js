import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import store from '../dummy-store';
import AddNote from './AddNote';

describe(`AddNote component`, () => {
    const props = {
      store: store
    }
  
    it('renders AddNote by default', () => {
      const wrapper = shallow(<AddNote />)
      expect(toJson(wrapper)).toMatchSnapshot()
    })
  
    it('renders form from store when in props', () => {
      const component = shallow(<AddNote {...props} />)
        .find('form')
      expect(toJson(component)).toMatchSnapshot()
    })
  })