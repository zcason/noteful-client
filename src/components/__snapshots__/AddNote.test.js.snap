// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNote component renders AddNote by default 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`AddNote component renders form from store when in props 1`] = `null`;
